W&B Training: Serverless RL launch

dbrian57 opened the pull request:
## Description
This adds documentation for the new W&B Training Serverless RL service. It includes:

* A general guide about the service
* An API reference
* A Redoc integration so that we can render the engineering team's OpenAPI spec


<!-- preview-links-comment -->
📄 **[View preview links for changed pages](https://github.com/wandb/docs/pull/1682#issuecomment-3340339826)**

dbrian57 committed:
Adds new directory for service

dbrian57 committed:
Adds WB training and serverless RL section

dbrian57 committed:
Adds Serverless RL reference section with redoc integration

dbrian57 committed:
Remove unnecessary files

dbrian57 committed:
Feedback via D. Corbitt

dbrian57 committed:
Updates Serverless RL API page with proper spec

dbrian57 committed:
Adds back the Training branding

dbrian57 committed:
Updates front matter with training branding

dbrian57 committed:
new identifier

dbrian57 committed:
use cases

dbrian57 committed:
fixes top nav rendering in cloudflare preview

dbrian57 committed:
Adds static light and dark logos

dbrian57 committed:
fixes top nav

dbrian57 committed:
CSS references assets directory now

dbrian57 committed:
new test implementation of redoc using redoc2

dbrian57 committed:
force light mode on redoc page

dbrian57 committed:
fix scrolling issue in redoc2

dbrian57 committed:
Replaces old layout with new one

dbrian57 committed:
Fixes width of redoc page

dbrian57 committed:
Updates API spec URL for Redoc

dbrian57 committed:
adds prereqs page

dbrian57 committed:
adds Serverless RL sub-section

dbrian57 committed:
Adds available models section

dbrian57 committed:
Adds use serverless RL section

dbrian57 committed:
Adds usage and limits section

dbrian57 committed:
Adds API ref placeholder section

arcticfly commented on content/en/guides/training/prerequisites.md:
```diff
@@ -0,0 +1,33 @@
+   Get your API key at [https://wandb.ai/authorize](https://wandb.ai/authorize)
+
+3. **A W&B project**  
+   Create a project in your W&B account to track usage
```
"to track usage" -> "to track usage, record training metrics, and save trained models"

dbrian57 committed:
Updates marketecture diagram

github-actions[bot] committed:
Optimised images with calibre/image-actions

dbrian57 committed:
Removes unnecessary static assets

dbrian57 committed:
Merge branch 'docs/training-serverless-rl' of https://github.com/wandb/docs into docs/training-serverless-rl

github-actions[bot] committed:
Optimised images with calibre/image-actions

dbrian57 committed:
remove unnecessary navbar changes

dbrian57 committed:
Merge branch 'docs/training-serverless-rl' of https://github.com/wandb/docs into docs/training-serverless-rl

dbrian57 committed:
adds back whitespace to navbar partial

github-actions[bot] committed:
Optimised images with calibre/image-actions

arcticfly commented on content/en/guides/training/serverless-rl/usage-limits.md:
```diff
@@ -0,0 +1,42 @@
+
+### Inference
+
+Trajectories are batched at each training step and, together with rewards from your reward function (often an LLM-as-judge), are used to update a training copy of the LoRA. The training workload runs on a separate GPU cluster within Serverless RL. During the public preview, training is free. Pricing for training will be announced at general availability (GA).
```
I think this content corresponds to the Training section, and we'll need to update it for inference.

dbrian57 committed:
Training descriptions

dbrian57 committed:
Merge branch 'docs/training-serverless-rl' of https://github.com/wandb/docs into docs/training-serverless-rl

dbrian57 committed:
Adds Training to homepage

dbrian57 committed:
new training icons

dbrian57 committed:
Updates marketecture from official Figma

github-actions[bot] committed:
Optimised images with calibre/image-actions

dbrian57 committed:
Adds public preview verbage

dbrian57 committed:
Merge branch 'docs/training-serverless-rl' of https://github.com/wandb/docs into docs/training-serverless-rl

dbrian57 committed:
Feedback

github-actions[bot] committed:
Optimised images with calibre/image-actions

arcticfly committed:
David edits (#1693)

Minor edits to content and a new "Using Trained Models" page.


<!-- preview-links-comment -->
📄 **[View preview links for changed
pages](https://github.com/wandb/docs/pull/1693#issuecomment-3374283089)**

arcticfly commented on content/en/guides/training/_index.md:
```diff
@@ -0,0 +1,14 @@
+weight: 5
+---
+
+Now win public preview, W&B Training offers serverless reinforcement learning (RL) for post-training large language models (LLMs) to improve their reliability performing multi-turn, agentic tasks while also increasing speed and reducing costs. 
```
Minor nit: "win" -> "in"

dbrian57 committed:
updates to David's copy and shuffles some stuff around

dbrian57 committed:
spell check

dbrian57 committed:
spelling error

arcticfly commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+
+This tells W&B Inference exactly where to find your model when initially loading it for inference.
+
+The following examples show you how to make inference requests to your trained model using a CuRL request or the [Python OpenAI SDK](https://openai.github.io/openai-agents-python/).
```
I actually don't think this will work with the agents SDK (https://openai.github.io/openai-agents-python/), but it should work with the normal OpenAI SDK (https://github.com/openai/openai-python).

mdlinville commented:
1. Something looks off about the image names and the imgrefs. I don't think the mode stuff should be there.
2. I don't know why I'm not getting a preview link for `guides/_index.md` but it now says that there are 3 major products and then goes on to list 4. https://docs-training-serverless-rl.docodile.pages.dev/guides/
3. Now, there are two separate TOC entries that both load up the quickstart. https://docs-training-serverless-rl.docodile.pages.dev/quickstart/ and https://docs-training-serverless-rl.docodile.pages.dev/guides/models_quickstart/
4. https://docs-training-serverless-rl.docodile.pages.dev/guides/models/ doesn't mention Inference outside of the marchitecture diagram.
5. https://docs-training-serverless-rl.docodile.pages.dev/guides/training/prerequisites/ is not ideal. Why not add the prereqs to https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/serverless-rl/ (note the duplicated URL components) and/or https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/using-trained-models/ 
6. Detailed pricing info belongs on https://wandb.ai/site/pricing/, probably not on https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/usage-limits/. Understood if the current situation is temporary.
7. It might be better to combine https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/available-models/ into https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/using-trained-models/ unless I'm misunderstanding the contents.

arcticfly commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+PROJECT = "my-project"
+
+client = OpenAI(
+    base_url="https://api.inference.coreweave.com/v1",
```
Oh, definitely meant to put this url: `https://api.training.wandb.ai/v1`

arcticfly commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+client = OpenAI(
+    base_url="https://api.inference.coreweave.com/v1",
+    api_key=WANDB_API_KEY,
+    default_headers={"OpenAI-Project": f"{ENTITY}/{PROJECT}"},
```
The user shouldn't actually need to include this, I added it by accident.

mdlinville commented:
> ~1. Something looks off about the image names and the imgrefs. I don't think the mode stuff should be there.~ Update: This is fine and works.
> 2. I don't know why I'm not getting a preview link for `guides/_index.md` but it now says that there are 3 major products and then goes on to list 4. https://docs-training-serverless-rl.docodile.pages.dev/guides/
> 3. Now, there are two separate TOC entries that both load up the quickstart. https://docs-training-serverless-rl.docodile.pages.dev/quickstart/ and https://docs-training-serverless-rl.docodile.pages.dev/guides/models_quickstart/
> 4. https://docs-training-serverless-rl.docodile.pages.dev/guides/models/ doesn't mention Inference outside of the marchitecture diagram.
> 5. https://docs-training-serverless-rl.docodile.pages.dev/guides/training/prerequisites/ is not ideal. Why not add the prereqs to https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/serverless-rl/ (note the duplicated URL components) and/or https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/using-trained-models/
> 6. Detailed pricing info belongs on https://wandb.ai/site/pricing/, probably not on https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/usage-limits/. Understood if the current situation is temporary.
> 7. It might be better to combine https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/available-models/ into https://docs-training-serverless-rl.docodile.pages.dev/guides/training/serverless-rl/using-trained-models/ unless I'm misunderstanding the contents.

mdlinville commented on content/en/guides/training/_index.md:
Worth defining RL here?

mdlinville commented on content/en/guides/training/_index.md:
```diff
@@ -0,0 +1,14 @@
+
+Now in public preview, W&B Training offers serverless reinforcement learning (RL) for post-training large language models (LLMs) to improve their reliability performing multi-turn, agentic tasks while also increasing speed and reducing costs. 
+
+W&B Training includes integration with [ART](https://art.openpipe.ai/getting-started/about), a flexible RL fine-tuning framework; [RULER](https://openpipe.ai/blog/ruler), a universal verifier; and a fully managed serverless RL backend on CoreWeave cloud, so you can run RL training loops without provisioning, configuring, or managing GPUs and infrastructure.
```
We try to avoid complex lists like this. It might be better as a list.

mdlinville commented on content/en/guides/training/serverless-rl/_index.md:
```diff
@@ -0,0 +1,40 @@
+
+Reinforcement learning (RL) is a set of powerful training techniques that you can use in many kinds of training setups, including on GPUs that you own or rent directly. Serverless RL can provide the following advantages in your RL post-training:
+
+* Lower training costs
```
These would be better as dictionary lists like:

```
* **Lower training costs**: By multiplexing....
```

mdlinville commented on content/en/guides/training/serverless-rl/_index.md:
```diff
@@ -0,0 +1,40 @@
+
+## How Serverless RL uses W&B services
+
+Serverless RL uses a combination of the following W&B services to operate:
```
I'm not sure we call these "services", maybe components?

ngrayluna commented on content/en/_index.md:
```diff
@@ -65,6 +65,30 @@ Use [W&B Inference]({{< relref "/guides/inference/" >}}) to access leading open-
+- [API Reference]({{< relref "/guides/training/api-reference/" >}})
+
+</div>{{% /card %}}
+{{< /cardpane >}}
```
Might need some CSS here. The width of the card stretches across (see attached image)
<img width="1727" height="980" alt="Screenshot 2025-10-06 at 4 44 28 PM" src="https://github.com/user-attachments/assets/e6ad4558-10e1-4c01-a5fd-0999307630b0" />

mdlinville commented on content/en/guides/training/serverless-rl/_index.md:
```diff
@@ -0,0 +1,40 @@
+* [Artifacts]({{< relref "guides/core/artifacts" >}}): To store and version the LoRA adapters
+* [Weave (optional)]({{< relref "guides/models" >}}): To gain observability into how the model responds at each step of the training loop
+
+Serverless RL is currently in public preview. We only charge for the use of inference and the storage of artifacts. We do not charge for adapter training during the preview period. We will make training pricing details available in the next few weeks.
```
```suggestion
Serverless RL is in public preview. During the preview, you are charged only for the use of inference and the storage of artifacts, during the preview period, not for adapter training. We will make training pricing details available in the next few weeks.
```
is there any way to avoid needing the future-facing language about pricing details?

dbrian57 replied:
Worked this in and I just deleted this future stuff.

ngrayluna commented on content/en/guides/training/api-reference.md:
```diff
@@ -0,0 +1,16 @@
+---
```
Wondering if we can integrate this content within the _index.md page (for example). The experience feels a little disjoint to see this page in the left TOC, but then ask users to go to the reference docs.

dbrian57 replied:
Totally agree

dbrian57 replied:
I fixed it so that it links directly! 🙌

mdlinville commented on content/en/guides/training/serverless-rl/available-models.md:
```diff
@@ -0,0 +1,17 @@
+  See the models you can train with Serverless RL.
+---
+
+Serverless RL currently only supports a single open-source foundation model for training. We will add more models soon.
```
Can we avoid or minimize the future-facing language?

dbrian57 replied:
Deleted!

mdlinville commented on content/en/guides/training/serverless-rl/available-models.md:
```diff
@@ -0,0 +1,17 @@
+
+Serverless RL currently only supports a single open-source foundation model for training. We will add more models soon.
+
+If there's a particular model you're interested in getting support for, email use at support@wandb.com.
```
```suggestion
To express interest in a particular model, contact [support](mailto:support@wandb.ai).
```

mdlinville commented on content/en/guides/training/serverless-rl/serverless-rl.md:
```diff
@@ -0,0 +1,9 @@
+---
+description: Directions on how to get started using Serverless RL.
```
```suggestion
description: Get started using Serverless RL.
```

dbrian57 replied:
This is almost exactly what I had before, but I removed it because it wasn't technically a link to a getting started guide :/

mdlinville commented on content/en/guides/training/serverless-rl/serverless-rl.md:
```diff
@@ -0,0 +1,9 @@
+weight: 10
+---
+
+Serverless RL is supported through [OpenPipe's ART framework](https://art.openpipe.ai/getting-started/about) and the W&B Training API. 
```
Link to API docs here instead of in line 9?

ngrayluna replied:
(Suggestion feel free to ignore) Since the content on this page is so short, we can add this as an H2 (for now) on the training/serverless-rl/_index.md page. I agree that the TOC navigation is appealing, though.

mdlinville commented on content/en/guides/training/serverless-rl/usage-limits.md:
```diff
@@ -0,0 +1,46 @@
+---
+title: "Usage Information and Limits"
```
We try to use sentence case for page titles, I believe.

mdlinville commented on content/en/guides/training/serverless-rl/usage-limits.md:
```diff
@@ -0,0 +1,46 @@
+By default, Serverless RL currently supports up to 2000 concurrent requests per user and 6000 per project. If you exceed your rate limit, the Inference API returns a `429 Concurrency limit reached for requests` response. To avoid this error, reduce the number of concurrent requests your training job or production workload makes at once. If you need a higher rate limit, you can request one at supp ...
+
+{{< alert title="Note" >}}
+## Personal entities unsupported
```
I'm not sure that we should have H2 headings in notes.

dbrian57 committed:
Reweights top level items in left-nav

mdlinville commented on content/en/guides/training/serverless-rl/usage-limits.md:
```diff
@@ -0,0 +1,46 @@
+
+## Geographic restrictions
+
+Serverless RL is only available in supported geographic locations. For more information, see the [Terms of Service](https://docs.coreweave.com/docs/policies/terms-of-service/terms-of-use#geographic-restrictions).
```
Just confirming that this is the correct page to link to? It doesn't mention Serverless RL or Training and it's a CoreWeave page, not a W&B page.

mdlinville commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+
+After training a model with Serverless RL, it is automatically deployed and ready to use in production. To make a request to the OpenAI-compatible endpoint, follow the instructions below.
+
+### Overview
```
Nit: This doesn't look like an overview.

ngrayluna commented on content/en/guides/training/prerequisites.md:
```diff
@@ -0,0 +1,33 @@
+
+3. **A W&B project**  
+   Create a project in your W&B account to track usage, record training metrics, and save trained models
+
```
```suggestion
## Sign up and create an API key
To authenticate your machine with W&B, you must first generate an API key at [wandb.ai/authorize](https://wandb.ai/authorize). Copy the API key and store it securely.

## Create a project in W&B

Create a project in your W&B account to track usage, record training metrics, and save trained models. See [Insert link to a quickstart or tutorial]

```

ngrayluna commented on content/en/guides/training/_index.md:
```diff
@@ -0,0 +1,14 @@
+
+W&B Training includes integration with [ART](https://art.openpipe.ai/getting-started/about), a flexible RL fine-tuning framework; [RULER](https://openpipe.ai/blog/ruler), a universal verifier; and a fully managed serverless RL backend on CoreWeave cloud, so you can run RL training loops without provisioning, configuring, or managing GPUs and infrastructure.
+
+Learn how to use W&B Training's [Serverless RL]({{< relref "/guides/training/serverless-rl" >}}) to post-train your models.
```
Maybe add the next first step? E.g. the prerequisites page?
```suggestion
To get started, satisfy the prerequisites described...
```

ngrayluna commented on content/en/guides/training/serverless-rl/_index.md:
```diff
@@ -0,0 +1,40 @@
+---
```
Dumb question: Do we plan to have more products with Training? If nay, consider removing nesting Serverless RL and instead moving the content up one level within `/training`

dbrian57 replied:
@ngrayluna Yep, there are gonna be more products for Training. I already tried to nix the nesting despite that and people were not having it.

dbrian57 committed:
Makes product layout 2x2 on front page

ngrayluna commented on content/en/guides/training/serverless-rl/_index.md:
```diff
@@ -0,0 +1,40 @@
+weight: 5
+---
+
+Now in public preview, Serverless RL helps developers post-train LLMs to learn new behaviors and improve reliability, speed, and costs when performing multi-turn agentic tasks. We provision the training infrastructure for you while allowing full flexibility in your environment's setup. Serverless RL gives you instant access to a managed training cluster that elastically auto-scales to dozens of G ...
```
Grep, Ctrl + F and replace "we" with "W&B"

ngrayluna replied:
Should we mention that we are using CW? And possibly link to their docs for more info on the hardware side?

ngrayluna commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+  Make inference requests to the models you've trained.
+---
+
+After training a model with Serverless RL, it is automatically deployed and ready to use in production. To make a request to the OpenAI-compatible endpoint, follow the instructions below.
```
What does this mean "After training a model with Serverless RL, it is automatically deployed and ready to use in production."? Where is it deployed to?

ngrayluna commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+
+When making inference requests, set your base URL to `https://api.training.wandb.ai/v1/` and use your W&B API key for authentication. Finally, set your model name to match the following schema:
+
+`wandb-artifact:///<entity>/<project>/<model-name>:<step>`
```
Wrap code in triple ```

e.g.

```bash
wandb-artifact:///<entity>/<project>/<model-name>:<step>
```

ngrayluna commented on content/en/guides/training/serverless-rl/using-trained-models.md:
```diff
@@ -0,0 +1,68 @@
+
+For example, if your W&B team is named `email-specialists`, your project is called `mail-search`, your trained model is named `agent-001`, and you wanted to deploy it on step 25 (presumably because that step did the best on your evals), you would set the model name in your chat completion requests to the following:
+
+`wandb-artifact:///email-specialists/mail-search/agent-001:step25`
```
Same applies here

dbrian57 committed:
Feedback via Noah, Matt, and David

dbrian57 committed:
Fixes API reference page

dbrian57 committed:
remove unnecessary menu frontmatter from reference

dbrian57 committed:
Adds coreweave link

dbrian57 committed:
rewrite inference doc

dbrian57 committed:
minor copy changes

dbrian57 committed:
renames file and corrects some grammar

dbrian57 committed:
renames file for real

mdlinville commented on content/en/guides/training/_index.md:
```diff
@@ -4,11 +4,15 @@ menu:
\ No newline at end of file
+* [ART](https://art.openpipe.ai/getting-started/about), a flexible RL fine-tuning framework.
+* [RULER](https://openpipe.ai/blog/ruler), a universal verifier. 
+* A fully-managed backend on CoreWeave cloud.
```
```suggestion
* A fully-managed backend on CoreWeave Cloud.
```
https://www.coreweave.com/contact-us

mdlinville commented on content/en/guides/training/serverless-rl/usage-limits.md:
```diff
@@ -29,17 +29,17 @@ Serverless RL stores checkpoints of your trained LoRAs so you can evaluate, serv
 
```
Maybe for a followup, this H2 could be Limitations, then you have the option to use H3s to further classify them without needing to stack admonitions.

mdlinville commented on content/en/guides/training/serverless-rl/use-trained-models.md:
```diff
@@ -0,0 +1,79 @@
+title: "Use your trained models"
+linkTitle: "Use trained models"
+weight: 20
+description: >
```
Total nit but you don't need a multi-line description here, to be consistent with the existing content.

mdlinville commented on content/en/guides/training/serverless-rl/use-trained-models.md:
````diff
@@ -0,0 +1,79 @@
+The model's endpoint uses the following schema:
+
+```
+wandb-artifact:///<entity>/<project>/<model-name>:<step>
````
If the entity isn't currently supported, does that change this?

ngrayluna commented on content/en/guides/training/prerequisites.md:
```diff
@@ -0,0 +1,33 @@
+   Get your API key at [https://wandb.ai/authorize](https://wandb.ai/authorize)
+
+3. **A W&B project**  
+   Create a project in your W&B account to track usage, record training metrics, and save trained models
```
Heads up that formatting looks a bit wonky here.

dbrian57 replied:
ah, weird. I could have sworn I copied your other suggestion over. Updating

dbrian57 committed:
Feedback from M. Linville and N. Luna

dbrian57 committed:
Updates ToS link and LoRA deletion link, fixes spelling error

dbrian57 committed:
Updates from launch meeting

dbrian57 committed:
link fix
